waypost-io/waypost
# Create CRON job for counting exposures in experiments

- Installed the `node-cron` library
- Created a file called `refreshExptStatus.js` inside a folder called `cron` inside of the `server` directory
- There are 3 main functions that get called:

1. `getActiveExperiments()` - The CRON job first fetches all active experiments by querying from `experiments WHERE date_ended IS NULL`, and returns the array of id's.
2. `countExposures()` - This takes in the output of the previous function, and queries their `experiments` table (using their query) to count how many users were added to each experiment on the previous day, and which treatment (true or false). It returns the row data with each row corresponding to one experiment-cohort.
3. If this succeeds, we run the next function. I don't have anything for the "else" case here but I'm thinking we should send an email or something letting them know it failed and that they need to run it again.
4. `updateExposures()` - This takes in the output of the previous function, and formats the data into a string which will go in the SQL INSERT statement after "VALUES", then runs the SQL statement.

# Testing
- Make sure your `user_event_data_dump.sql` file has the word `treatment` instead of `variation_id` in lines 68 and 222.
- Make sure you have experiment_id 1 running (date_ended is not null)
- Run the data dump, then do an UPDATE on the `experiments` table to set the timestamp to yesterday (`'2022-03-15'`) since that's the date that the CRON job will filter for.
- Run the script by opening your terminal in the `server` directory, and run `node ./cron/refreshExptStatus.js`. For testing purposes this will only run once, but when we deploy it we'll use the CRON scheduling code that I've included in the comments at the bottom of the file.

# Notes
- I haven't added the part for calculating means for metrics yet, wanted to get this out and reviewed first before I do that.
- I had to make some changes to a few other `db` files to get this to work

# Questions
- As I mentioned in step 3, should we send some kind of notification if there's a failure and the data didn't get updated?
- It doesn't handle the case where we're missing data from days before yesterday (like if it ever fails), since it only queries for yesterday. Should we change this so it counts the cumulative number of users from the experiment start date up to the previous day so that we don't have to worry about missing data?
